omp/solver/batch_bicgstab_kernels.cpp

```diff
@@ -0,0 +1,215 @@
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/solver/batch_bicgstab_kernels.hpp"
+
+#include <omp.h>
+#include "omp/base/config.hpp"
+// include device kernels for every matrix and preconditioner type
+#include "omp/matrix/batch_struct.hpp"
+#include "reference/log/batch_logger.hpp"
+#include "reference/matrix/batch_csr_kernels.hpp"
+#include "reference/matrix/batch_dense_kernels.hpp"
+#include "reference/preconditioner/batch_identity.hpp"
+#include "reference/preconditioner/batch_jacobi.hpp"
+#include "reference/stop/batch_criteria.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace omp {
+
+
+/**
+ * @brief The batch Bicgstab solver namespace.
+ *
+ * @ingroup batch_bicgstab
+ */
+namespace batch_bicgstab {
+
+
+namespace {
+
+
+using gko::kernels::reference::advanced_spmv_kernel;
+using gko::kernels::reference::batch_scale;
+using gko::kernels::reference::spmv_kernel;
+namespace batch_dense = gko::kernels::reference::batch_dense;
+using gko::kernels::reference::BatchIdentity;
+using gko::kernels::reference::BatchJacobi;
+namespace stop = gko::kernels::reference::stop;
+namespace batch_log = gko::kernels::reference::batch_log;
+
+
+#include "reference/solver/batch_bicgstab_kernels.hpp.inc"
+
+
+}  // unnamed namespace
+
+
+template <typename T>
+using BatchBicgstabOptions =
+    gko::kernels::batch_bicgstab::BatchBicgstabOptions<T>;
+
+template <typename StopType, typename PrecType, typename LogType,
+          typename BatchMatrixType, typename ValueType>
+static void apply_impl(
+    std::shared_ptr<const OmpExecutor> exec,
+    const BatchBicgstabOptions<remove_complex<ValueType>> &opts, LogType logger,
+    PrecType prec, const BatchMatrixType &a,
+    const gko::batch_dense::UniformBatch<const ValueType> &left,
+    const gko::batch_dense::UniformBatch<const ValueType> &right,
+    const gko::batch_dense::UniformBatch<ValueType> &b,
+    const gko::batch_dense::UniformBatch<ValueType> &x)
+{
+    using real_type = typename gko::remove_complex<ValueType>;
+    const size_type nbatch = a.num_batch;
+    const auto nrows = a.num_rows;
+    const auto nrhs = b.num_rhs;
+
+    GKO_ASSERT(batch_config<ValueType>::max_num_rhs >=
+               nrhs);  // required for static allocation in stopping criterion
+
+    const int local_size_bytes =
+        gko::kernels::batch_bicgstab::local_memory_requirement<ValueType>(
+            nrows, nrhs) +
+        PrecType::dynamic_work_size(nrows, a.num_nnz) * sizeof(ValueType);
+    using byte = unsigned char;
+
+    Array<byte> local_space(exec, local_size_bytes);
+
+#pragma omp parallel for firstprivate(logger) firstprivate(local_space)
```

Comment:
I hope the `firstprivate` here will not do a deep copy of the `local_space` for each thread.